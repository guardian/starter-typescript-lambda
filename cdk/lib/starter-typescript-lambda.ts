import { GuScheduledLambda } from '@guardian/cdk';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { Duration } from 'aws-cdk-lib';
import type { App } from 'aws-cdk-lib';
import { Schedule } from 'aws-cdk-lib/aws-events';
import { Runtime } from 'aws-cdk-lib/aws-lambda';

const APP_NAME = 'starter-typescript-lambda';

export enum LoggingLevel {
	SILLY,
	TRACE,
	DEBUG,
	INFO,
	WARN,
	ERROR,
	FATAL,
}

export class StarterTypescriptLambda extends GuStack {
	constructor(
		scope: App,
		id: string,
		props: GuStackProps,
		loggingLevel: number = LoggingLevel.WARN,
	) {
		super(scope, id, props);

		const otherConfig = {
			app: APP_NAME,
			runtime: Runtime.NODEJS_18_X,
			fileName: `${APP_NAME}.zip`,
			timeout: Duration.millis(45000),
			environment: {
				Bucket: `${APP_NAME}-dist`,
				Stage: this.stage,
				LoggingLevel: loggingLevel.toString(),
			},
		};

		new GuScheduledLambda(this, APP_NAME, {
			handler: 'dist/lambda/index.handler',
			rules: [
				{
					schedule: Schedule.cron({ hour: '10', minute: '00', weekDay: '2' }),
				},
			],
			monitoringConfiguration: {
				noMonitoring: true,
			},
			...otherConfig,
		});
	}
}
