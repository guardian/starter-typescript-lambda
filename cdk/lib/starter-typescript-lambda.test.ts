import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { StarterTypescriptLambda } from "./starter-typescript-lambda";

describe("The StarterTypescriptLambda stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new StarterTypescriptLambda(app, "StarterTypescriptLambda", { stack: "playground", stage: "TEST" });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
