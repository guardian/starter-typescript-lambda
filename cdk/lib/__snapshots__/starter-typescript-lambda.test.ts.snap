// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The StarterTypescriptLambda stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "startertypescriptlambda16F2E652": {
      "DependsOn": [
        "startertypescriptlambdaServiceRoleDefaultPolicy454B579F",
        "startertypescriptlambdaServiceRole02688246",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "playground/TEST/starter-typescript-lambda/starter-typescript-lambda.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "starter-typescript-lambda",
            "Bucket": "starter-typescript-lambda-dist",
            "LoggingLevel": "4",
            "STACK": "playground",
            "STAGE": "TEST",
            "Stage": "TEST",
          },
        },
        "Handler": "dist/lambda/index.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "startertypescriptlambdaServiceRole02688246",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "starter-typescript-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/starter-typescript-lambda",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 45,
      },
      "Type": "AWS::Lambda::Function",
    },
    "startertypescriptlambdaServiceRole02688246": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "starter-typescript-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/starter-typescript-lambda",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "startertypescriptlambdaServiceRoleDefaultPolicy454B579F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/playground/TEST/starter-typescript-lambda/starter-typescript-lambda.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/playground/starter-typescript-lambda",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/playground/starter-typescript-lambda/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "startertypescriptlambdaServiceRoleDefaultPolicy454B579F",
        "Roles": [
          {
            "Ref": "startertypescriptlambdaServiceRole02688246",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "startertypescriptlambdastartertypescriptlambdacron001020AllowEventRuleStarterTypescriptLambdastartertypescriptlambda35AFBAF7B1B92612": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "startertypescriptlambda16F2E652",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "startertypescriptlambdastartertypescriptlambdacron001020C3F1D1BC",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "startertypescriptlambdastartertypescriptlambdacron001020C3F1D1BC": {
      "Properties": {
        "ScheduleExpression": "cron(00 10 ? * 2 *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "startertypescriptlambda16F2E652",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
  },
}
`;
